packages/loopring_v3/contracts/core/impl/libtransactions/TransferTransaction.sol

```diff
@@ -56,12 +62,17 @@ library TransferTransaction
     {
         // Read the transfer
         Transfer memory transfer = readTransfer(data, offset);
+        TransferAuxiliaryData memory auxData = abi.decode(auxiliaryData, (TransferAuxiliaryData));
+
+        // Check validUntil
+        require(block.timestamp < auxData.validUntil, "WITHDRAWAL_EXPIRED");
```

Comment:
Actually I have another idea that may work:

1. Conver the root into a Root Node which contains some global layer2 state variables, such as the smallest and largest `validUntil` values. Then on L2, we only need to check two values against block.timestamp. 

2. `validUntil` should compare with `inputTimestamp`, not with 'block.timestamp'.

Reply:
> Actually I have another idea that may work:
> 
>     1. Conver the root into a Root Node which contains some global layer2 state variables, such as the smallest and largest `validUntil` values. Then on L2, we only need to check two values against block.timestamp.

But we'd still have to check that `minValidUntil <= tx.validUntil <= maxValidUntil` onchain for each conditional tx, and do all that work to just do `minValidUntil < inputTimestamp < maxValidUntil` on layer 2 which would also be practically free on layer 1? Not sure I understand the idea. The simple time check is very cheap either way (50 gas).

>     2. The offchain validation of `validUntil` for EdDSA signed off-chain transactions is done by comparing it with the `inputTimestamp` of the block, so the onchain validation of `validUntil` should also  compare with `inputTimestamp`, not with 'block.timestamp'.

I guess that's right, but why not use the more accurate time when we can, `inputTimestamp` is only there because we can't use 'block.timestamp' in the circuit. Fine with me either way. :)

Reply:
I prefer to use `inputTimestamp` as there will be fewer failures and make the logic consistent.